### DISCUSSION TITLE
Discussion: potential merge with `pytest-copie`

### DISCUSSION THREAD

**Discussion by kenibrewer** [noirbizarre/pytest-copier#10] [General]
Key discussion points. 

1. What are the features that are present in `pytest-copier` that would be worth bringing over to `pytest-copie` (or vice-versa)?
2. Under who's leadership should the combined project be led?
3. Is @noirbizarre willing to donate the `pytest-copier` name to a merged project?
4. Is `pytest-copier` currently being used in production anywhere despite being explicitly labeled as a PoC?

**Comment by noirbizarre** [comment]
Hi @kenibrewer 👋🏼 

Thanks for reaching me. I will answer your points but let me bring a bit of context before: I started this project from scratch for my needs and those from the company I work for or with. It started as an experimentation (hence the lack of doc and tests for the moment) at a time when there was nothing existing. We are actively using it and it was meant from day one to be donated to @copier-org (if they accept it) as soon as it is fully working for our cases. You know the drill: as soon as a personal project become a community driven one, some bad practice can't persist (and this kind of try and publish dev is not a good practice, I know, but it allows to quickly get something up and working). Right now, I am finalizing it as it is only missing the snapshot part (and tests and docs obviously).

So given this context:

1. I need to dig into `pytest-copie` to answer (but from the first sight at the readme, all features provided by `pytest-copie` are already in `pytest-copier` except template path change), but here's the list of `pytest-copier` features:
  - expose the current copier template as a project typed fixture (it is meant for testing a single copier template)
  - allow copy and update
  - allows testing rendered context values without rendering the project
  - allows to tests rendered answer
  - allows to test the entire rendered projects
  - allows to easily commands in the rendered project
  - WIP: snapshoting based on `syrupy` (I have it working, just finalizing the matchers and filters part. I already proposed to syrupy the path/directory snapshoting part: https://github.com/tophat/syrupy/issues/829)

2 and 3 are the same to me: it was my plan from day 1 to give the project to the @copier-org as soon as I have a "fully" working solution. It is still the case. TBH, in the past, I burnt-out because of overload of open source project to maintain. I don't want to go back to it, and I am looking to contribute to existing and living projects (aka. with one or more existing maintainers and ideally a working organization). I believe single maintainers projects are not sustainable on long term. 
As a consequence, I'm willing to:
- give it as it is to copier-org and let them maintain it
- give it to copier-org and be part of maintainers
- merge it with yours to have the best of both world
I am OK with any, as long as there is a shared maintenance and leadership and we can agree to keep some backward-compatibility in the future devs.

4. Yes it is. On my side on non public (yet) personal projects. At my work too, we have some Python projects using copier managed templates tested with this lib. It's working just fine, but without the snapshoting part, updating the template and the tests is a long process (hence the fact I want to finish this part before giving the project). 

I was on holidays, so I haven't had time to finish yet, but it's in my current high priority todolist so ~expect a delivery to copier-org this month (I would love to be able to do it this week, but I'll avoid false promises I won't be able to fill)~ (Sorry I mixed with the other issue/question)

In the meantime, I'll take a look at `pytest-copie` to see where it overlaps, where it doesn't.

Also, If you want to try the project and you have some question, don't hesitate. Same if you have some specific use cases and want to know if they are supported...

**Comment by noirbizarre** [comment]
I converted the issue as a discussion

**Comment by noirbizarre** [comment]
OK, so I read `pytest-copie` code in details and read the doc, have good night sleep and here's my thought:

## Similarities

The core feature is almost the same: feed the fixture with a template and allow running `copier copy`.

But except from that, we have a lot of differences, some that can be merged to benefit from both, some are different design choices where we should decide which one we keep.

## Mergeable features

### Template from command line parameter vs template from fixture

I think we can have both. We just need the fixture to default to the command line parameter and the command line parameter to default to the config root (this is my fixture default). This way:
- default to the project root as template
- user can use the parameter to change the template root
- user can override the fixture to change the template root

### Git isolation and tagging

`pytest-copier` make a temporary copy of the template to isolate it like `pytest-copie` do, but it also handles tagging of this copy to remove some `copier` warnings and be able to then test update between 2 tags and provide `gitconfig` isolation to avoid user personal settings to mess with the test (or prevent them running if you use GPG signing as example).

### Syntactic sugar

Lowering the entry cost to make it the more accessible possible for non-devs was a goal (we are testing copier template, not some code and some colleagues working on them are not devs), so I added some helpers to simplify test code:
- a basic `Project.run()` helper allowing to run commands in the rendered project root
- the `copy()` and `update()` commands accept data as `kwargs` to avoid writing dicts
- the resulting `CopierProject` implement a `Path-like` interface allowing `project / "some.file"` syntax

### Other helpers

In `pytest-copier`, the `copier` fixture also provide the following helpers:

### `copier_defaults` fixture

Allows to provide some extra default values to avoid providing some recurrent parameters.

### `copier_template_paths` fixture

A whitelist of paths to consider as "the template". Currently mandatory, but I was planning to change that to have both a `copier_template_include` and a `copier_template_exclude` optional fixture with `fn_match` glob patterns to refine the selection if wanted.

### `CopierFixture.context()` method

Allows testing the rendering context given some answers (including defaults). Useful when you just want to some answer default using template filter or some data generated by extensions without going through the whole rendering.

### `CopierFixture.setenv()` and `CopierFixture.delenv()` methods

Given `Copier` use `plumbum` internally, `monkeypatch.{set,del}env()` changes are ignored and not seen by copier. Those methods allow to both monkeypatch the built-in `os.environ` and the `plumbum.local.env` used internally by `Copier`. Helpful if your template depends on some environment variables.

### Tests

Same as documentation, `pytest-copie` have tests while `pytest-copier` don't. Obviously, given the API proximity, they can easily be merged and adapted.

### Typing

Both packages are typed, but `pytest-copie` is adding stubs while `pytest-copier` only rely on inline typing.

I might be missing something there, but I think stubs are redundant with the inline typing.

Also, `pytest-copie` still uses some pre-3.8 syntax for some typing while I try to benefit from `from __future__ import annotations`. Given I use the `pyupgrade` `ruff` plugin, they would be upgraded automatically.

## Design choices

### Result-based vs Exception-based

`pytest-copie` is using an intermediate `Result` object while `pytest-copier` raise exception while failing.
I intentionally chose to have this approach for the following reason:
- raising exception allows to style and control `pytest` output
- when developing a plugin or a copier extension, it gives a proper stacktrace prettified by `pytest` in case of error in `copier`
- it allows catching some expected exceptions to cut the stacktrace when needed (example: avoid having a subprocess error on `tasks` execution and instead have a proper task execution error)
- avoid the need to assert for result_code: if it fails, exceptions are just properly handled as test failure by `pytest`
- avoid an intermediate object (i.e. `result.project` vs `project`)

### Assertions helpers

The `assert_equal`, `assert_answers` and the snapshotting features in `pytest-copier` are very opinionated. In fact, after thinking a lot about it, they can be removed or moved into an `integrations` sub package.

### `pdm` vs `hatchling`

This is very personal, but I contribute to `pdm` and it's now my daily driver for all Python projects.
To be honest, I am both maintainer on `pdm` and `commitizen` (for which I also provide a template for changelogs) so I am not neutral on this question. But my main argument for `pdm` would be built-in scripts: I find it cleaner to have all housekeeping script listed somewhere and easily modifiable and on community driven and open source project I think it lower the entry cost to have all dev commands discoverable with `pdm run -l`.

### `tox` vs `pdm`

I used both, and I got back to `tox`, mostly because of [tox-pdm](https://github.com/pdm-project/tox-pdm) which allows having a single source of trust for both scripts and dependencies.

### Documentation

`pytest-copie` have one, `pytest-copier` don't.

`pytest-copie` is using Google/Napoleon docstrings while in `pytest-copier` I was intending to use sphinx syntax. I am OK with Google syntax with typing 👍🏼

However, the project is himself fed by a template of my own which is not yet published, and I was intending to add the doc part (which is part of my template). Target was not the same: `mkdocs`+`mkdocs-material`. Same for README, Markdown vs restructured text. I tend to drop `restructuredtext` as much as possible as I believe it is raising entry cost compared to Markdown, so I will push for Markdown if possible.

## Leadership

The more it's shared, the better it is to me. Review based PR merging has proven efficient for this.

For the namespace holding it: I already have a working `pytest-copier` repository, PyPI package and I can create the associated readthedocs.
My goal is to give the package to @copier-org if this is OK for them but until it's done, if we merge, I propose we use this repository and which I can give extra permissions to any maintainer.
If @copier-org is OK, we move this repository to their org.

Does it seem OK for you ?

## Proposal

Would you be OK if I do merge proposal as a PR on this repository ? This way, you can review it and tell me if this is OK and acceptable to you, and we can iterate together on it. I think having both documentation in parallel might help spot differences too.

So, what do you think about it ? I'd like to read your vision too on those 4 questions and on those answers.

**Comment by kenibrewer** [comment]
Thanks for your very detailed response @noirbizarre. To clarify, I'm not affiliated with `pytest-copie`, I'm planning to build an open-source best practices copier template (similar to https://github.com/fpgmaas/cookiecutter-poetry) and was looking for a solution for testing. I had been planning to use `pytest-copie` simply because it has documentation and tests. However, I also didn't realize how far along your version is. 

My main goal here was just facilitating a discussion between you and @12rambau to figure out a unified strategy forward. I didn't want to start using one of your two packages only to "bet on the wrong horse" and need to convert all my testing to the package with better community support down the line. 

As an external user, I really like your suggestion about there being a group of maintainers who collaborate together. I also like the suggestion of preparing to donate the project to @copier-org .

From my end, I can start building the `copier-poetry` template that I've been envisioning. My vision actually sounds very similar to the copier template that you were planning on having this project descend from. That could provide a shortcut for `pytest-copier` to get the documentation that it currently needs. Hopefully can you and @12rambau can decide on a path forward so I can know what I should be using for testing. :)

As an aside, the snapshotting based on `syrupy` is very exciting. That definitely seems like a super useful feature.

**Comment by 12rambau** [comment]
Sorry for the delay in my follow-up but I was a bit overwhelmed by work related stuff, so less time to dedicate to side-projects. 

## Context 

For starters be aware that my first thought when I moved from `cookiecutter` to `copier` was to use this plugin thus the issue #9. As there was some delay in your answers, no documentations and no apparent links to other repositories I guessed (accurately aparently) that it was a company driven repository that you were generously openly developing in GitHub. I had a terrible experience working on this type of repositories, where outside-of-company PR and issues are either completely disregarded or simply left drying until they get completely out of sync with main. My last misadventure was with `ipyvuetify` where the maintainers comes every 3 month to do their stuff and that's all not trusting their community for anything. 

Afraid I thus decided to write my own thing with the open-source community in mind: 
- Build the documentation on the fly 
- Advertise the repository on [copier-org](https://github.com/orgs/copier-org/discussions/1324)
- Let external users contribute (A huge thanks to @GenevieveBuckley)
- publish the package in conda-forge as well as pypi
- Use it in the open-source communities I work in (Jupyter, gee-community, sphinx-contrib, openForis)
- Of course my long term goals is/was to handhover the repo to `copier-org` the same way I gave away many of my libs to `sphinx-contrib`, `gee-community` or `openForis` in the past. It's more robust and give proper visibility to the code. 


If you are willing to integrate some of this in your repository and hand it hover to to copier-org community I'm all in for a merge. If the need of your company will remain the principal driver of the development, I'm out I would find it undelicate to block an open-source name for a private package (even though I know GitHub offers 0 solution to create a private registry https://github.com/github/roadmap/issues/94)

It seems you are more going for the first option so Let's dive into it. 

## Design choice 

### Result-based vs Exception-based

I decided to go with the Result-based technique because I am mirroring the widely used `pytest-cookie` lib. It's robust as people have not complained there and more importantly it will facilitate the transition. It's already a bummer to move from `cruft`/`cookiecutter` to copier so if I wanted to ligntend this burden.

Also I designed some templates that are heavily linked to external/commercial API so I need the result object to run extra checks on the generated repo. Is it possible in your implementation ? I have the feeling I could use `project` and if you can confirm it I have no strong opinion on this one. 

### Assertions helpers

I think they are fine but might not be useful for my use cases. 

### pdm VS hatchling

Before I was using `setuptools` and I think Python packaging is not great in general compared to other languages.

I usually follow the pyOpenSci graph to choose:  
![package selection graph](https://raw.githubusercontent.com/pyOpenSci/python-package-guide/ddb511edb659c723787a96cc9cc9b21e26d7297e/images/python-package-tools-decision-tree.svg)

And of course we are falling in the "Any tool!" area. 
  
I would like to use something that is most likely to stay or at least get close to pypa if they decide to converge to something [eventually](https://discuss.python.org/t/python-packaging-strategy-discussion-part-1/22420/147?page=8). `hatch` being hosted by pypa itself I think it's a better candidate and I like their implementation.

copier itself is using poetry so I guess we all disagree here. 

`commitizen` is my go to tool since I discovered it so no conflict here. 

### pdm vs nox 

Look carefully I'm actually not using `tox` but `nox` :smile: which is doing a similar thing but is configured with a python file. I find easier to maintain and discover. I don't like to have scripts in the `pyproject.toml` file as it's a "configuration" file. Thus I don't want to use hatch or any other builder scripts as it's mixing config with workflows. I'm open to discussion but very opinionated on this one :wink:.

### Documentation

I'm a `Sphinx` contributor and maintainer of many `Sphinx` related libs (`pydata-sphinx-theme`, `sphinx-book-theme`, etc...) so I'm very partial on this subject. 

Even though .rst is more complex to handle, .md is nowhere near the level of customization you can get from a sphinx based documentation. The fact that I have already created one (that is perfectly aligned with copier-org look and feel) I would like to keep my implementation.

If the problem you forsee is writting the rst files themselve, there is a very powerfull extention for sphinx to write everything in markdown and keep the extensibility of sphinx called [myst](https://github.com/executablebooks/MyST-Parser)

for the readme you can use .md it doesnt have any consequences for the doc or published pypi display. 

## Leadership 

To move forward with the merge, I think the first steps are to create both documentation and tests for your package. I am willing to do it as I have experience in both things. I would simply need more rights to the repositories (maintainers so i can create branch without forking and make review) and admin rights on readthedocs site to fasten the set up. 

Once we are happy with the merging I think we can use the conversation I opened on copier-org to add it and request to remain admin of the repo. 

## Summary 

I think every point could be converted into a dedicated issue. 

### next steps 

- create automated tests 
- crete documentation 
- set up github actions
- set up coverage report

### open questions

- which build tool to use ? (I fine by keeping `pdm` for now I don't really care as long as we get a wheel)
- where to set up the automated scripts ? (pdm vs nox)
- which documentation tool to use ? (strong opinion over Sphinx on my side, I hate mkdocs)

**Reply by noirbizarre** [reply]
> Sorry for the delay in my follow-up but I was a bit overwhelmed by work related stuff, so less time to dedicate to side-projects.
> ## Context
> 
> For starters be aware that my first thought when I moved from `cookiecutter` to `copier` was to use this plugin thus the issue #9. As there was some delay in your answers, no documentations and no apparent links to other repositories I guessed (accurately aparently) that it was a company driven repository that you were generously openly developing in GitHub. I had a terrible experience working on this type of repositories, where outside-of-company PR and issues are either completely disregarded or simply left drying until they get completely out of sync with main. My last misadventure was with `ipyvuetify` where the maintainers comes every 3 month to do their stuff and that's all not trusting their community for anything.
> 
> Afraid I thus decided to write my own thing with the open-source community in mind:
> 
>     * Build the documentation on the fly
> 
>     * Advertise the repository on [copier-org](https://github.com/orgs/copier-org/discussions/1324)
> 
>     * Let external users contribute (A huge thanks to @GenevieveBuckley)
> 
>     * publish the package in conda-forge as well as pypi
> 
>     * Use it in the open-source communities I work in (Jupyter, gee-community, sphinx-contrib, openForis)
> 
>     * Of course my long term goals is/was to handhover the repo to `copier-org` the same way I gave away many of my libs to `sphinx-contrib`, `gee-community` or `openForis` in the past. It's more robust and give proper visibility to the code.
> 
> 
> If you are willing to integrate some of this in your repository and hand it hover to to copier-org community I'm all in for a merge. If the need of your company will remain the principal driver of the development, I'm out I would find it undelicate to block an open-source name for a private package (even though I know GitHub offers 0 solution to create a private registry [github/roadmap#94](https://github.com/github/roadmap/issues/94))
>
> It seems you are more going for the first option so Let's dive into it.

Just to reassure you first: it's not company driven. It's a tool I made for my needs first and the company I currently work for benefit from the initiative. If it was the case, I would have published it under the company account. It was meant for donation from day 1, the issue with doc and tests is mostly by lack of time on my side (I contribute to or maintain a lot of projects).

> ## Design choice
> ### Result-based vs Exception-based
> 
> I decided to go with the Result-based technique because I am mirroring the widely used `pytest-cookie` lib. It's robust as people have not complained there and more importantly it will facilitate the transition. It's already a bummer to move from `cruft`/`cookiecutter` to copier so if I wanted to ligntend this burden.
> 
> Also I designed some templates that are heavily linked to external/commercial API so I need the result object to run extra checks on the generated repo. Is it possible in your implementation ? I have the feeling I could use `project` and if you can confirm it I have no strong opinion on this one.

As of today, the only thing you won't have is the `exit_code` but given it's not standardised on `copier` I don't think this is an issue. Problem with the result based approach:
- good for testing copier itself, not bringing extra info or facilities for testing copier templates
- all your test will start with `assert result.exit_code == 0` and if this assertion fail, you won´t  benefit from `pytest` assertion rewriting, you won't have any details on why did it fail OR you will have too much non-related details if relying only on the builtin stdout/stderr capture as you will have `copier` output mixed with the subprocess error stacktrace
- it prevent to use any of the `pytest` hooks relying on exception types
- I don't see a case where you try to render a template but you expect a failure with a specific `exit_code`

However, if you need to rely on `exit_code` for whatever reason, given it is wrapped in a dedicated exception, it totally possible to expose it in the exception (it's not yet the case because I didn't need it but I will add it ASAP) allowing this kind of test:

```python
def test_exit_code(copier):
    with pytest.raises(CopierError) as ei:
        copier.copy()
    assert ei.exception.exit_code == 0
```

Every other field from the result object is available:
- `result.project` is the returned value itself (a `CopierProject` instance)
- `result.answers` is a `CopierProject` attribute
- `result.exception` is directly raised except for tasks execution where it is cleaned up (because resulting in a subprocess error stacktrace by default with obfuscated/broken output)
- `result.exit_code` wil be in the exception if any

> ### Assertions helpers
> 
> I think they are fine but might not be useful for my use cases.

👍🏼

I still thinks after usage that except some integrated helpers relying on `pytest-copier` specifics (like `syrupy` snapshots), most of the assertions can be removed as they bring extra dependencies and kind of forces you into some tooling.

> ### pdm VS hatchling
> 
> Before I was using `setuptools` and I think Python packaging is not great in general compared to other languages.
> 
> I usually follow the pyOpenSci graph to choose: ![package selection graph](https://raw.githubusercontent.com/pyOpenSci/python-package-guide/ddb511edb659c723787a96cc9cc9b21e26d7297e/images/python-package-tools-decision-tree.svg)
> 
> And of course we are falling in the "Any tool!" area.

I didn't know this decision graph, it's interesting but to me it misses a lot of things: those tool aim to be far more than just a package manager. As soon as you integrate questioning about housekeeping, community handling, phases outside the packaging... we go back to square one for most of them:
- most of the time you will need some extra tooling for the scripts, `pdm` has a powerfull builtin scripts engine (TLDR: which I know well given I wrote most of it) allowing to have `pyproject.toml` as a single source of trust, not requiring contributors to have another tool to execute tasks like `invoke` or `poe`
- this where I got back from `nox` to `tox`: with `tox-pdm`, no more double declaration: the tox file is just an interface to pdm: install this group of dependencies and run this script


> I would like to use something that is most likely to stay or at least get close to pypa if they decide to converge to something [eventually](https://discuss.python.org/t/python-packaging-strategy-discussion-part-1/22420/147?page=8). `hatch` being hosted by pypa itself I think it's a better candidate and I like their implementation.

Interesting, Frostming is `pdm` author and core dev on poetry. He is one of the contributor on all the `pyproject.toml` related PEPs on which `pdm` is literally cited.
This discussion is interesting because, following this post they are literally discussing how to do X and Y while X and Y are already in pdm.
And if you fear that later `hatch` become the Python tool, `pdm` nailed PEP517 and as of today is the only tool allowing to work with any PEP517 backend while being itself a PEP517 compliant tool (so works with `pypa/build`). It allows using `hatch` PEP517 for a smooth transition if necessary.

However, I think that while `hatch` provide lots of feature for big projects, I find it overly complicated for most projects while `pdm` provides all the same features just simpler. And hatch does not supports scripts, so perfect when you want a dedicated tool for that, but I think overkill for a 2 module package like this project.

> copier itself is using poetry so I guess we all disagree here.

There's one details on this: `pdm` was made by a `poetry` core devs (Frostming) and receive contributions from poetry devs which acted the tech debt (mostly the fact that the tool did use it's own formalism at the time official PEP on the topic were released, especially PEP621) was too much to be able to deliver something sticking with standards PEP on short or midterm. So `pdm` is made with the `poetry` experience and we literaly speak and work with the some of the `poetry` team which is currently tackling this topic. `poe` is also planned for integration in poetry, but from the guy which will do it (he is present on our PDM Discord), it's planned, not started.

> 
> `commitizen` is my go to tool since I discovered it so no conflict here.

👍🏼 

> ### pdm vs nox
> 
> Look carefully I'm actually not using `tox` but `nox` 😄 which is doing a similar thing but is configured with a python file. I find easier to maintain and discover. I don't like to have scripts in the `pyproject.toml` file as it's a "configuration" file. Thus I don't want to use hatch or any other builder scripts as it's mixing config with workflows. I'm open to discussion but very opinionated on this one 😉.

cf. above. I know you are using `nox`, I have been using it too but now with `pdm` it ~feel~ is an unnecessary extra maintenance burden as it duplicate scripts with less benefits in my opinion. The alternative to duplication with `nox` is `nox`-only scripts and housekeeping which got us back to square one: 2 tools, 2 source of thrust to manage in parallel, longer onboarding of newcomers.

> ### Documentation
> 
> I'm a `Sphinx` contributor and maintainer of many `Sphinx` related libs (`pydata-sphinx-theme`, `sphinx-book-theme`, etc...) so I'm very partial on this subject.
> 
> Even though .rst is more complex to handle, .md is nowhere near the level of customization you can get from a sphinx based documentation. The fact that I have already created one (that is perfectly aligned with copier-org look and feel) I would like to keep my implementation.
> 
> If the problem you forsee is writting the rst files themselve, there is a very powerfull extention for sphinx to write everything in markdown and keep the extensibility of sphinx called [myst](https://github.com/executablebooks/MyST-Parser)
> 
> for the readme you can use .md it doesnt have any consequences for the doc or published pypi display.

I have a too strong opinion on this. While Sphinx was a good tool to me at start, at mostly have been a source of problems in professional envirnement where people have always opposed me those (legit) arguments:
- `sphinx` is a python-dev only focused tool and restructuredtext mostly Python world only skill which makes hard for non Python dev to use it (tech writers, QA, product owners... just don't want to learn a new syntax) with such a specific scope
- there is not a single tool for non tech people, only text editors integrations in best case scenarios while Markdown have editors for all sensitivity
- there are almost no platform integration at all while most platforms support Markdown builtin
- rest/sphinx integrations are rest/sphinx only and specific meaning that you are lucky if there is a sphinxcontrib extension for the thing you want, but most of the time you only have PyMarkdown extension/integrations
- given all tooling is dev oriented, there is no way for a non dev to edit with live preview with a Python env

So I had to drop sphinx and the extensions I had written. However, if I remember (I need to check) `copier` itself is using `sphinx` so for consistency I am OK to use `sphinx`, especially if intending to give them the project.

> ## Leadership
> 
> To move forward with the merge, I think the first steps are to create both documentation and tests for your package. I am willing to do it as I have experience in both things. I would simply need more rights to the repositories (maintainers so i can create branch without forking and make review) and admin rights on readthedocs site to fasten the set up.
> 
> Once we are happy with the merging I think we can use the conversation I opened on copier-org to add it and request to remain admin of the repo.

I am totally OK with that. Yes, that was my intent to give you maintainer permissions if we go this way.

> ## Summary
> 
> I think every point could be converted into a dedicated issue.
> ### next steps
> 
>     * create automated tests

Yep, it should be fairly easy given the base you have in tests
 
>     * crete documentation

I think you mostly did the job with yours which will serve as a base with the merge

>     * set up github actions
> 
>     * set up coverage report

It´s already there: there is a full CI including coverage report, automated release... The CI part is just missing the tests in fact.

> ### open questions
> 
>     * which build tool to use ? (I fine by keeping `pdm` for now I don't really care as long as we get a wheel)

So let's keep `pdm`. Plus I would be happy if it permits you to discover the tool and reconciliate you with Python packaging and project management (it wea my case, I switched to `pdm` and contributed to it at a time I was so angry to the state of packaing/project management in Python I started to write my own tool)!


>     * where to set up the automated scripts ? (pdm vs nox)

In `pytest-copier`, this is already done, scripts are in `pyproject.toml:tool.pdm.scripts`, `tox` just reuse them and as said above, CI too 

>     * which documentation tool to use ? (strong opinion over Sphinx on my side, I hate mkdocs)

As said in the dedicated section, I have the opposite position, been using `mkdocs`+`mkdocs-material` for the last 10 years I think and not going back (mostly because I can't use the tool and my skills in pro environements).

But as I said, I am OK to use `sphinx` for concistency with `copier` itself. Plus, you already did the job AND I don´t think there is much documentation to write.


Anyway, I will preapre that in the coming days 👍🏼

**Comment by 12rambau** [comment]
I'll try to reconcile with Python packaging and I encourage you to have a look to myst to try to reconcile with Sphinx :wink:, bonus this syntax is supported in notebooks in jupyterlab.

**Reply by noirbizarre** [reply]
Oh I know `myst` and I like it (commonmark has been stagning for too long and it is good to have such an initiative trying to formalize an extended version with extensions support), I even wrote a pure markdownit implementation to be able to use it with pelican without pulling `docutils`.
The main issue I have with `myst` is basically the same, yes it provides markdown support to sphinx but:
- the implementation was basically empty (no directives or helpers) outside of sphinx (I reimplemented them all, just not yet published)
- to maintain some compatibility, some choices in the specs were discutables and not logical outside of sphinx 
- the build (need to check if this is still the case) was a bit dodgy for the `docutils`-free distribution (which by the way means that it's not thought as a standard then something which can be integrated in sphinx but as sphinx extension first. If it was the case, it would have a standard docutils-free install and the sphinx part would be an extra)
- specs were moving too much by that time, and it was hard to follow without being in the sphinx team (same, it must have changed, I hope)
- specs were really hard to use outside sphinx. There was no separation between what is `myst` and what is `myst-sphinx` integration making hard to test it using the specs like you would on commonmark or gfm specs

But I didn't said I don't know sphinx and its ecosystem. I used it intensively for a long time. I just said that it's was not fitted for corporate environments I was evolving into, where you don't do just Python and that I had to adapt.

Migration to `mkdocs` was a bit hard at start (because it was young and lacking some feature that I used in sphinx). But today, I don't regret to have made the move: there is nothing anymore that I miss from sphinx and situation change for the opposite: easier to find tooling for `mkdocs` than for sphinx. And for the opensource or personal projects, it is really a limiter to doc-only contributions. However, I know that sphinx is tailored for documentation of big projects and as such provide tooling that mkdocs won't ever. But given I never had to work on a project requiring it (except Python itself or Celery maybe, but I didn't contributed to the docs), it's clearly not my first choice anymore.

**Edit:** my `myst` tool is [mystit](https://pypi.org/project/mystit/) but it's a placeholder as I never published it because the day I had something working, there was a major change published on the specs and given I had no way to efficiently identify what changed, I put it on hold until specs were stabilized. I will get back to it at some point)

**Comment by astrojuanlu** [comment]
Arrived here while looking for info on whether to adopt pytest-copier or pytest-copie. For now I think I'll toss a coin (I'm between pytest-copier just because I like PDM more, and pytest-copie because I'm migrating from pytest-cookie) but looking forward to see convergence here.

**Comment by simone-gaiarin** [comment]
> 4\. Is `pytest-copier` currently being used in production anywhere despite being explicitly labeled as a PoC?

In our company we are using it: https://github.com/flowtaleai/copier-python-poetry/

**Comment by kenibrewer** [comment]
Hey @12rambau  and @noirbizarre ! Wanted to check in about whether or not a merge is still being planned. I see this package as incredibly valuable, but am a bit too deep in other open-source commitments to help out here myself.

**Reply by 12rambau** [reply]
Same here so I left mine where it is as I have too much on my plate these days.

**Comment by looztra** [comment]
Ok, we are close to 2 years after the last message in this discussion and no merge happened (but hey, I'am also a maintainer of some small pypi packages and I know what it means to be busy somewhere else).

I started very recently to write some copier template tests and here is what I did:

- start with pytest-copier (because I liked the "copier.context()" part) and failed because of some non compatibility issues with latest copier version it seems (needs to be double checked)
- use pytest-copie
- and backport some pytest-copier fixtures to get something like "copier.context()" stuff working
- and copy cat the template preparation from pytest-copier (the pre-launch git commands) to save half of the test execution time (seems like copier spends a lot of time in i/o operations when the template is in dirty state)

So what's next?

- I can live for some time with my custom setup until a merge target is clarified, and I definitely contribute to the merge and the final solution if it happens.

But if "for some time" last too long:

- I'll be honest, if no news from any of the maintainers of pytest-copier (cc @noirbizarre) and pytest-copie (cc @12rambau), I may start a brand new pytest copier plugin (and scratch my head to find a nice name).
- I don't have an opinion right now about the Result-based vs Exception-based topic, I'm wondering if we can do both actually (did not spend any time right now to think about it)
- Regarding `pdm` vs `hatch` or something else, I would definitely go for `uv` (with `hatchling` for the packaging part), like I did on [yamkix](https://github.com/looztra/yamkix) for instance
- I would go for mkdocs (with Material) for the doc part (or maybe start directly with [zensical](https://zensical.org/))

yes, I know it sounds like the famous xkcd comics

**Reply by looztra** [reply]
by the way, the [copier](https://github.com/copier-org/copier) project did the switch to `uv` and `mkdocs`

**Reply by looztra** [reply]
> I don't have an opinion right now about the Result-based vs Exception-based topic, I'm wondering if we can do both actually (did not spend any time right now to think about it)

I read the thread more carefully, and @noirbizarre did provide a target to tackle the `Result-based vs Exception-based` topic

**Reply by 12rambau** [reply]
I'm exactly in the position you described at the start of your message, I have little time to dedicate to this project so I'm not going to code anything in the forseeable future but as I'm working on github again, I'll see the notifications much faster and I'll be more than happy to review PRs.
